conformance-test/v4SignedUrl.ts

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright 2019 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as assert from 'assert';
+import * as fs from 'fs';
+import * as path from 'path';
+
+import {Storage} from '../src/';
+
+const testCases = JSON.parse(fs.readFileSync(
+    path.join(
+        __dirname,
+        '../../conformance-test/test-data/v4SignedUrl.json',
+        ),
+    'utf-8'));
+
+const SERVICE_ACCOUNT = path.join(
+    __dirname, '../../conformance-test/fixtures/signing-service-account.json');
+
+describe('v4 signed url', () => {
+  let storage: Storage;
+
+  before(() => {
+    storage = new Storage({keyFilename: SERVICE_ACCOUNT});
+  });
+
+  // tslint:disable-next-line:no-any
+  testCases.forEach((testCase: any) => {
```

Comment:
With this pattern, when a test fails - you're going to get a stack trace that points to code in a loop.  How are we going to know specifically which code is failing?  I don't have a great answer, but I personally really dislike debugging these kinds of tests-in-a-loopsie-loop thingies.  Wondering if @bcoe has an answer.

Reply:
I've never loved this pattern in tests (auto generating the description feels pretty clever, and makes it hard to grep the codebase for a specific test case).

I might instead look into using a snapshot pattern, it's really similar but addresses some of @JustinBeckwith's concerns, and I think is a really cool pattern to have in our tool-belt. It works like this:

You have a unit test with an assertion that looks like this:


*test.js*

```js
  it('reports coverage for script that exits normally', () => {
    const { output } = runSomeOperation()
    output.should.matchSnapshot()
  })
```

You have a snapshot file that maps the assertion names to captured outputs:

*test.js.snapshot*

```js
exports[`c8 check-coverage allows --check-coverage when executing script 1`] = "{
  some
}"`;
```

You tend to pair this pattern with some process for generation, and you run the snapshots in generation mode when changes have been made -- in your case this might be changes to a fixture.

---

Here's the library I've been using to add this functionality to `mocha`, I really love it:

https://www.npmjs.com/package/chai-jest-snapshot